lib/email_api/template_handler.py

```diff
@@ -0,0 +1,170 @@
+import os
+from typing import Dict, Optional
+
+from yaml import safe_load, YAMLError
+from jinja2 import Environment, FileSystemLoader, Template
+from jinja2.exceptions import TemplateError, TemplateNotFound
+from exceptions.email_template_error import EmailTemplateError
+
+
+class TemplateHandler:
+    """
+    TemplateHandler class is used to get template files and render them using jinja2 templating engine.
+    An email body is a string made up of one or more rendered jinja2 templates.
+    """
+
+    # Holds absolute filepath to email template metadata (from current dir)
+    # .../st2-cloud-pack/lib/email_api/email_templates.yaml
+    EMAIL_TEMPLATE_METADATA_FP = os.path.normpath(
+        os.path.join(
+            os.path.dirname(os.path.realpath(__file__)), "./email_templates.yaml"
+        )
+    )
+
+    # Holds absolute dirpath to directory where email template files are stores
+    # .../st2-cloud-pack/email_templates
+    EMAIL_TEMPLATE_ROOT_DIR = os.path.normpath(
+        os.path.join(
+            os.path.dirname(os.path.realpath(__file__)), "../../email_templates"
+        )
+    )
+
+    def __init__(self):
+        """constructor class"""
+        self._template_env = Environment(
+            loader=FileSystemLoader(self.EMAIL_TEMPLATE_ROOT_DIR)
+        )
+        self._template_metadata = self._load_all_metadata()
+
+    def _load_all_metadata(self) -> Dict:
+        """
+        Method which reads in email templates yaml file. This file holds metadata for each template name.
+        """
+        with open(self.EMAIL_TEMPLATE_METADATA_FP, "r", encoding="utf-8") as stream:
+            try:
+                return safe_load(stream)
+            except YAMLError as exc:
+                raise EmailTemplateError(
+                    "could not load template metadata file"
+                ) from exc
+
+    @staticmethod
+    def _parse_template_attrs(
+        template_name: str, template_schema: Dict[str, str], given_vals: Dict[str, str]
+    ) -> Dict[str, str]:
```

Comment:
Ditto about the types here, there's a lot of pairs of dicts and strings floating that imply to me we have correlated types that can be tied into a struct